src/vs/platform/terminal/common/capabilities/commandDetection/promptInputModel.ts

```diff
@@ -280,50 +280,45 @@ export class PromptInputModel extends Disposable implements IPromptInputModel {
 
 		const buffer = this._xterm.buffer.active;
 		let line = buffer.getLine(commandStartY);
+		const absoluteCursorY = buffer.baseY + buffer.cursorY;
+		let cursorIndex: number | undefined;
+
 		let commandLine = line?.translateToString(true, this._commandStartX);
-		if (!line || commandLine === undefined) {
-			if (this._shellType === PosixShellType.Fish) {
-				commandStartY += 1;
-				line = buffer.getLine(commandStartY);
-				if (line) {
-					commandLine = line.translateToString(true);
-				}
-				if (!commandLine || !line) {
-					return;
-				}
-			} else {
-				this._logService.trace(`PromptInputModel#_sync: no line`);
+		if (this._shellType === PosixShellType.Fish && (!line || !commandLine)) {
+			commandStartY += 1;
+			line = buffer.getLine(commandStartY);
+			if (line) {
+				commandLine = line.translateToString(true);
+				cursorIndex = absoluteCursorY === commandStartY ? buffer.cursorX : commandLine?.trimEnd().length;
+			}
+			if (commandLine === undefined || !line) {
 				return;
 			}
+		} else if (line === undefined || commandLine === undefined) {
+			this._logService.trace(`PromptInputModel#_sync: no line`);
+			return;
 		}
```

Comment:
I think this can fall through to simplify things?

```ts
			if (line) {
				commandLine = line.translateToString(true);
				cursorIndex = absoluteCursorY === commandStartY ? buffer.cursorX : commandLine?.trimEnd().length;
			}
		}
		if (line === undefined || commandLine === undefined) {
			this._logService.trace(`PromptInputModel#_sync: no line`);
			return;
		}
```